PolymorphicDispatcher#compare violates the contract for Java Comparators

Sometimes, we get the following exception when our Xsemantics-generated type system is initialized:

```
1) Error injecting constructor, java.lang.IllegalArgumentException: Comparison method violates its general contract!
  at eu.numberfour.n4js.typesystem.CustomInternalTypeSystem.<init>(Unknown Source)
  at eu.numberfour.n4js.typesystem.CustomInternalTypeSystem.class(Unknown Source)
  while locating eu.numberfour.n4js.typesystem.CustomInternalTypeSystem
  while locating eu.numberfour.n4js.xsemantics.InternalTypeSystem
    for field at eu.numberfour.n4js.typesystem.N4JSTypeSystem.ts_internal(Unknown Source)
  at org.eclipse.xtext.service.MethodBasedModule.configure(MethodBasedModule.java:57)
  while locating eu.numberfour.n4js.typesystem.N4JSTypeSystem
    for field at eu.numberfour.n4js.validation.AbstractN4JSDeclarativeValidator.ts(Unknown Source)
  at org.eclipse.xtext.service.MethodBasedModule.configure(MethodBasedModule.java:57)
  while locating eu.numberfour.n4js.validation.validators.N4JSProjectSetupValidator
Caused by: java.lang.IllegalArgumentException: Comparison method violates its general contract!
	at java.util.TimSort.mergeHi(TimSort.java:899)
	at java.util.TimSort.mergeAt(TimSort.java:516)
	at java.util.TimSort.mergeForceCollapse(TimSort.java:457)
	at java.util.TimSort.sort(TimSort.java:254)
	at java.util.Arrays.sort(Arrays.java:1512)
	at java.util.ArrayList.sort(ArrayList.java:1454)
	at java.util.Collections.sort(Collections.java:175)
	at org.eclipse.xtext.util.PolymorphicDispatcher.getDeclaredMethodsOrderedBySpecificParameterType(PolymorphicDispatcher.java:353)
	at org.eclipse.xtext.util.PolymorphicDispatcher.<init>(PolymorphicDispatcher.java:167)
	at org.eclipse.xtext.util.PolymorphicDispatcher.<init>(PolymorphicDispatcher.java:160)
	at it.xsemantics.runtime.XsemanticsRuntimeSystem$2.<init>(XsemanticsRuntimeSystem.java:86)
	at it.xsemantics.runtime.XsemanticsRuntimeSystem.buildPolymorphicDispatcher1(XsemanticsRuntimeSystem.java:86)
	at eu.numberfour.n4js.xsemantics.InternalTypeSystem.init(InternalTypeSystem.java:535)
	at eu.numberfour.n4js.xsemantics.InternalTypeSystem.<init>(InternalTypeSystem.java:531)
	at eu.numberfour.n4js.typesystem.CustomInternalTypeSystem.<init>(CustomInternalTypeSystem.java:39)
	at sun.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
	at sun.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:62)
	at sun.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
	at java.lang.reflect.Constructor.newInstance(Constructor.java:423)
	at com.google.inject.internal.DefaultConstructionProxyFactory$1.newInstance(DefaultConstructionProxyFactory.java:85)
	at com.google.inject.internal.ConstructorInjector.construct(ConstructorInjector.java:85)
	at com.google.inject.internal.ConstructorBindingImpl$Factory.get(ConstructorBindingImpl.java:254)
	at com.google.inject.internal.ProviderToInternalFactoryAdapter$1.call(ProviderToInternalFactoryAdapter.java:46)
	at com.google.inject.internal.InjectorImpl.callInContext(InjectorImpl.java:1031)
	at com.google.inject.internal.ProviderToInternalFactoryAdapter.get(ProviderToInternalFactoryAdapter.java:40)
	at com.google.inject.Scopes$1$1.get(Scopes.java:65)
	at com.google.inject.internal.InternalFactoryToProviderAdapter.get(InternalFactoryToProviderAdapter.java:40)
	at com.google.inject.internal.FactoryProxy.get(FactoryProxy.java:54)
	at com.google.inject.internal.SingleFieldInjector.inject(SingleFieldInjector.java:53)
	at com.google.inject.internal.MembersInjectorImpl.injectMembers(MembersInjectorImpl.java:110)
	at com.google.inject.internal.ConstructorInjector.construct(ConstructorInjector.java:94)
	at com.google.inject.internal.ConstructorBindingImpl$Factory.get(ConstructorBindingImpl.java:254)
	at com.google.inject.internal.ProviderToInternalFactoryAdapter$1.call(ProviderToInternalFactoryAdapter.java:46)
	at com.google.inject.internal.InjectorImpl.callInContext(InjectorImpl.java:1031)
	at com.google.inject.internal.ProviderToInternalFactoryAdapter.get(ProviderToInternalFactoryAdapter.java:40)
	at com.google.inject.Scopes$1$1.get(Scopes.java:65)
	at com.google.inject.internal.InternalFactoryToProviderAdapter.get(InternalFactoryToProviderAdapter.java:40)
	at com.google.inject.internal.SingleFieldInjector.inject(SingleFieldInjector.java:53)
	at com.google.inject.internal.MembersInjectorImpl.injectMembers(MembersInjectorImpl.java:110)
	at com.google.inject.internal.ConstructorInjector.construct(ConstructorInjector.java:94)
	at com.google.inject.internal.ConstructorBindingImpl$Factory.get(ConstructorBindingImpl.java:254)
	at com.google.inject.internal.ProviderToInternalFactoryAdapter$1.call(ProviderToInternalFactoryAdapter.java:46)
	at com.google.inject.internal.InjectorImpl.callInContext(InjectorImpl.java:1031)
	at com.google.inject.internal.ProviderToInternalFactoryAdapter.get(ProviderToInternalFactoryAdapter.java:40)
	at com.google.inject.Scopes$1$1.get(Scopes.java:65)
	at com.google.inject.internal.InternalFactoryToProviderAdapter.get(InternalFactoryToProviderAdapter.java:40)
	at com.google.inject.internal.InternalInjectorCreator$1.call(InternalInjectorCreator.java:204)
	at com.google.inject.internal.InternalInjectorCreator$1.call(InternalInjectorCreator.java:198)
	at com.google.inject.internal.InjectorImpl.callInContext(InjectorImpl.java:1024)
	at com.google.inject.internal.InternalInjectorCreator.loadEagerSingletons(InternalInjectorCreator.java:198)
	at com.google.inject.internal.InternalInjectorCreator.injectDynamically(InternalInjectorCreator.java:179)
	at com.google.inject.internal.InternalInjectorCreator.build(InternalInjectorCreator.java:109)
	at com.google.inject.Guice.createInjector(Guice.java:95)
	at com.google.inject.Guice.createInjector(Guice.java:72)
	at com.google.inject.Guice.createInjector(Guice.java:62)
	at eu.numberfour.n4js.N4JSInjectorProvider$1.createInjector(N4JSInjectorProvider.java:89)
	at eu.numberfour.n4js.N4JSStandaloneSetup.createInjectorAndDoEMFRegistration(N4JSStandaloneSetup.java:80)
	at eu.numberfour.n4js.N4JSInjectorProvider.internalCreateInjector(N4JSInjectorProvider.java:91)
	at eu.numberfour.n4js.N4JSInjectorProvider.getInjector(N4JSInjectorProvider.java:76)
	at eu.numberfour.n4js.N4JSInjectorProvider.setupRegistry(N4JSInjectorProvider.java:101)
	at org.eclipse.xtext.junit4.XtextRunner.methodBlock(XtextRunner.java:43)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:78)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:57)
	at org.junit.runners.ParentRunner$3.run(ParentRunner.java:290)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:71)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:288)
	at org.junit.runners.ParentRunner.access$000(ParentRunner.java:58)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:268)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:363)
	at org.junit.runners.Suite.runChild(Suite.java:128)
	at org.junit.runners.Suite.runChild(Suite.java:27)
	at org.junit.runners.ParentRunner$3.run(ParentRunner.java:290)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:71)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:288)
	at org.junit.runners.ParentRunner.access$000(ParentRunner.java:58)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:268)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:363)
	at org.junit.runner.JUnitCore.run(JUnitCore.java:137)
	at org.junit.runner.JUnitCore.run(JUnitCore.java:115)
	at org.apache.maven.surefire.junitcore.JUnitCoreWrapper.createRequestAndRun(JUnitCoreWrapper.java:113)
	at org.apache.maven.surefire.junitcore.JUnitCoreWrapper.executeEager(JUnitCoreWrapper.java:85)
	at org.apache.maven.surefire.junitcore.JUnitCoreWrapper.execute(JUnitCoreWrapper.java:54)
	at org.apache.maven.surefire.junitcore.JUnitCoreProvider.invoke(JUnitCoreProvider.java:134)
	at org.apache.maven.surefire.booter.ForkedBooter.invokeProviderInSameClassLoader(ForkedBooter.java:200)
	at org.apache.maven.surefire.booter.ForkedBooter.runSuitesInProcess(ForkedBooter.java:153)
	at org.apache.maven.surefire.booter.ForkedBooter.main(ForkedBooter.java:103)
```

Upon further investigation, it turns out that Xtext's `PolymorphicDispatcher`, upon initialization, tries to sort all dispatch methods using a compare method that violates the general contract for Java Comparators. That contract specifies that a compare method should define a total order, i.e., it should be antisymmetric, transitive, and total.

However, there are pairs of methods `(a,b)` where the last requirement is violated. Such methods are incomparable to each other, that is, neither is `a <= b` nor `a <= b`. The compare method returns `0` for such cases, thereby indicating that `a == b`, which is obviously wrong. See the following example:

```
public class PolymorphicDispatcherTest extends PolymorphicDispatcher<Object> {

	public PolymorphicDispatcherTest() {
		super(Collections.singletonList(new Object()), method -> false);
	}

	private Method findMethod(Class<?> current, String name, Class<?>... parameterTypes) {
		if (current == Object.class)
			throw new IllegalArgumentException("Cannot find method");
		try {
			return current.getDeclaredMethod(name, parameterTypes);
		} catch (java.lang.NoSuchMethodException | SecurityException e) {
			return findMethod(current.getSuperclass(), name, parameterTypes);
		}
	}

        private static class S {}

	private static class A extends C {
	}

	private static class B extends S {
	}

	private static class C extends S {
	}

	public void test(A a) {
	}

	public void test(B b) {
	}

	public void test(C c) {
	}

	public void showTransitivityViolation() {
		MethodDesc m1 = createMethodDesc(this, findMethod(getClass(), "test", A.class));
		MethodDesc m2 = createMethodDesc(this, findMethod(getClass(), "test", B.class));
		MethodDesc m3 = createMethodDesc(this, findMethod(getClass(), "test", C.class));

		int m1m2 = PolymorphicDispatcherTest.this.compare(m1, m2); // 0
		int m2m3 = PolymorphicDispatcherTest.this.compare(m2, m3); // 0
		int m1m3 = PolymorphicDispatcherTest.this.compare(m1, m3); // 1

		if (m1m2 <= 0 && m2m3 <= 0 && !(m1m3 <= 0))
			throw new IllegalStateException("Transitivity violation");
	}

	public static void main(String[] args) {
		PolymorphicDispatcherTest test = new PolymorphicDispatcherTest();
		test.showTransitivityViolation();
	}
}
```

In this example, the method returns `0` when comparing method pairs `m1,m2` and `m2,m3`, even though in both cases, the methods are not identical! Comparing `m1` and `m3` then yields `1`, which violates transitivity.

We think the problem is due to the fact that the subtype relation, on which the compare method is based, only defines a partial order. The compare method attempts to extend this to a total order in an incorrect way.

This exception is thrown only sometimes, and it has popped up only recently (to our knowledge). We think that this is due to the fact that the initial order of the methods to be sorted depends on the initial order of the methods returned by `Class#getDeclaredMethods`, which is undefined by Java. Furthermore, Java uses different sorting algorithms depending on the number of elements to be sorted. In our particular cases, we are trying to sort 59 methods.

--- a/org.eclipse.xtext.tests/src/org/eclipse/xtext/util/PolymorphicDispatcherTest.java
+++ b/org.eclipse.xtext.tests/src/org/eclipse/xtext/util/PolymorphicDispatcherTest.java
@@ -15,10 +15,14 @@ import java.lang.annotation.RetentionPolicy;
 import java.lang.annotation.Target;
 import java.lang.reflect.Method;
 import java.math.BigInteger;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
 
 import org.junit.Assert;
 import org.junit.Test;
 
+import com.google.common.base.Joiner;
 import com.google.common.base.Predicate;
 import com.google.common.collect.Lists;
 
@@ -182,6 +186,42 @@ public class PolymorphicDispatcherTest extends Assert {
 		} catch (IllegalStateException e) {
 			// ignore
 		}
+	}
+	
+	interface A {}
+	interface B {}
+	interface CB extends B {}
+	class DCBA implements A, CB, E {}
+	interface E {}
+	
+	@Test public void testAmbiguous2() throws Exception {
+		Object o1 = new Object() {
+			String label(A i) {
+				return "A";
+			}
+			String label(B b) {
+				return "B";
+			}
+			String label(CB cb) {
+				return "CB";
+			}
+			String label(E e) {
+				return "E";
+			}
+		};
+		PolymorphicDispatcher<String> dispatcher = new PolymorphicDispatcher<String>("label", Lists.newArrayList(o1)) {
+			@Override
+			protected String handleAmbigousMethods(List<MethodDesc> result, Object... params) {
+				List<String> p = new ArrayList<>();
+				for(MethodDesc md:result) {
+					p.add(md.getParameterTypes()[0].getSimpleName());
+				}
+				Collections.sort(p);
+				return Joiner.on(", ").join(p);
+			}
+		};
+		String string = dispatcher.invoke(new DCBA());
+		assertEquals("A, CB, E", string);
 	}
 
 	@Test public void testNullParams() throws Exception {

--- a/org.eclipse.xtext.util/src/org/eclipse/xtext/util/PolymorphicDispatcher.java
+++ b/org.eclipse.xtext.util/src/org/eclipse/xtext/util/PolymorphicDispatcher.java
@@ -15,7 +15,6 @@ import java.lang.reflect.Method;
 import java.util.ArrayList;
 import java.util.Arrays;
 import java.util.Collections;
-import java.util.Comparator;
 import java.util.Iterator;
 import java.util.List;
 
@@ -34,7 +33,7 @@ public class PolymorphicDispatcher<RT> {
 	private final List<? extends Object> targets;
 	private final Predicate<Method> methodFilter;
 
-	private List<MethodDesc> declaredMethodsOrderedBySpecificParameterType;
+	private List<MethodDesc> methods;
 
 	public static class DefaultErrorHandler<RT> implements ErrorHandler<RT> {
 		@Override
@@ -164,7 +163,7 @@ public class PolymorphicDispatcher<RT> {
 		this.targets = targets;
 		this.methodFilter = methodFilter;
 		this.handler = handler;
-		declaredMethodsOrderedBySpecificParameterType = getDeclaredMethodsOrderedBySpecificParameterType();
+		this.methods = getCandidateMethods();
 	}
 
 	protected class MethodDesc {
@@ -210,20 +209,21 @@ public class PolymorphicDispatcher<RT> {
 		}
 	}
 
+	/**
+	 * returns > 0 when o1 is more specific than o2,
+	 * 
+	 * returns == 0 when o1 and o2 are equal or unrelated,
+	 * 
+	 * returns < 0 when o2 is more specific than o1,
+	 */
 	protected int compare(MethodDesc o1, MethodDesc o2) {
-		final List<Class<?>> paramTypes1 = Arrays.asList(o1.getParameterTypes());
-		final List<Class<?>> paramTypes2 = Arrays.asList(o2.getParameterTypes());
-
-		// sort by number of parameters
-		if (paramTypes1.size() > paramTypes2.size())
-			return 1;
-		if (paramTypes2.size() > paramTypes1.size())
-			return -1;
+		final Class<?>[] paramTypes1 = o1.getParameterTypes();
+		final Class<?>[] paramTypes2 = o2.getParameterTypes();
 
 		// sort by parameter types from left to right
-		for (int i = 0; i < paramTypes1.size(); i++) {
-			final Class<?> class1 = paramTypes1.get(i);
-			final Class<?> class2 = paramTypes2.get(i);
+		for (int i = 0; i < paramTypes1.length; i++) {
+			final Class<?> class1 = paramTypes1[i];
+			final Class<?> class2 = paramTypes2[i];
 
 			if (class1.equals(class2))
 				continue;
@@ -251,21 +251,27 @@ public class PolymorphicDispatcher<RT> {
 			new Function<List<Class<?>>, List<MethodDesc>>() {
 				@Override
 				public List<MethodDesc> apply(List<Class<?>> paramTypes) {
+					// 'result' contains all best-matched MethodDesc for which 
+					// pairwise compare(m1, m2) == 0, meaning they're equal or unrelated. 
 					List<MethodDesc> result = new ArrayList<MethodDesc>();
-					Iterator<MethodDesc> iterator = declaredMethodsOrderedBySpecificParameterType.iterator();
-					while (iterator.hasNext()) {
+					Iterator<MethodDesc> iterator = methods.iterator();
+					NEXT: while (iterator.hasNext()) {
 						MethodDesc methodDesc = iterator.next();
 						if (methodDesc.isInvokeable(paramTypes)) {
 							if (result.isEmpty()) {
 								result.add(methodDesc);
 							} else {
-								int compare = compare(result.get(0), methodDesc);
-								if (compare < 0) {
-									result.clear();
-									result.add(methodDesc);
-								} else if (compare == 0) {
-									result.add(methodDesc);
+								Iterator<MethodDesc> it = result.iterator();
+								while(it.hasNext()) {
+									MethodDesc next = it.next();
+									int compare = compare(next, methodDesc);
+									if (compare < 0) {
+										it.remove();
+									} else if (compare > 0) {
+										continue NEXT;
+									}
 								}
+								result.add(methodDesc);
 							}
 						}
 					}
@@ -336,27 +342,21 @@ public class PolymorphicDispatcher<RT> {
 		return Void.class;
 	}
 
-	private List<MethodDesc> getDeclaredMethodsOrderedBySpecificParameterType() {
-		ArrayList<MethodDesc> cachedDescriptors = new ArrayList<MethodDesc>();
+	private List<MethodDesc> getCandidateMethods() {
+		ArrayList<MethodDesc> result = new ArrayList<MethodDesc>();
 		for (Object target : targets) {
 			Class<?> current = target.getClass();
 			while (current != Object.class) {
 				Method[] methods = current.getDeclaredMethods();
 				for (Method method : methods) {
 					if (methodFilter.apply(method)) {
-						cachedDescriptors.add(createMethodDesc(target, method));
+						result.add(createMethodDesc(target, method));
 					}
 				}
 				current = current.getSuperclass();
 			}
 		}
-		Collections.sort(cachedDescriptors, new Comparator<MethodDesc>() {
-			@Override
-			public int compare(MethodDesc o1, MethodDesc o2) {
-				return PolymorphicDispatcher.this.compare(o1, o2);
-			}
-		});
-		return cachedDescriptors;
+		return result;
 	}
 	
 	protected MethodDesc createMethodDesc(Object target, Method method) {

